cpp/src/arrow/array/validate.cc

```diff
@@ -637,6 +638,23 @@ struct ValidateArrayFullImpl {
 
 ARROW_EXPORT
 Status ValidateArrayFull(const ArrayData& data) {
+  if (data.null_count != -1) {
+    int64_t actual_null_count;
+    if (HasValidityBitmap(data.type->id()) && data.buffers[0]) {
+      // Do not call GetNullCount() as it would also set the `null_count` member
+      actual_null_count =
+          data.length - CountSetBits(data.buffers[0]->data(), data.offset, data.length);
+    } else if (data.type->id() == Type::NA) {
+      actual_null_count = data.length;
```

Comment:
This is only enforced by `arrow::NullArray::SetData`; it's possible that `data.null_count != data.length` which should be an error for `Type::NA`. Going a step further, we might want to just always require that `data.null_count == data.length` for `Type::NA` instead of also allowing `kUnknownNullCount`

Reply:
Disallowing `kUnknownNullCount` for `Type::NA` seems to fail on some compute tests. I'm not sure it's worth fixing: it might also make some code pointlessly more complex.

I don't really understand the rest of your comment. `data.null_count` is checked for incorrect values for `Type::NA` as well.

Reply:
Ah, nevermind: when I stop reading diagonally I see that we are ultimately checking `data.null_count == data.length`.

Separate note: this is also enforced by `ValidateArray`, so I suppose this check is redundant

Reply:
https://issues.apache.org/jira/browse/ARROW-13589 may help spot such redundant checks.